src/modules/hardware-wallet/src/LedgerApp.ts

```diff
@@ -0,0 +1,221 @@
+// https://github.com/radixdlt/radixdlt-ledger-app/blob/improve/change_cosmos_to_radix/docs/APDUSPEC.md
+
+import { ReturnCode, Instruction, CLA, AppState } from './types'
+import { sendApduMsg } from './HardwareWallet'
+import { cborByteOffsets } from './atomByteOffsetMetadata'
+import { Subject, Observable } from 'rxjs'
+import { RadixAtom, RadixTransferrableTokensParticle, RadixSpin, RadixECSignature, RadixBytes, RadixAddress } from 'radixdlt'
+
+const CHUNK_SIZE = 255
+
+let isSigning = false
+
+let version: string
+let versionResolve
+
+const versionPromise = new Promise((resolve, reject) => {
+    versionResolve = resolve
+})
+
+const observable = new Observable<AppState>(subscriber => {
+    let connected = false
+    setInterval(async () => {
+        if (isSigning) { return }
+        try {
+            const result = await getVersion()
+            if (!version) {
+                version = result.version
+                versionResolve()
+            }
+            if (connected) { return }
+        } catch (e) {
+            if (!connected) { return }
+        }
+        connected = !connected
+        subscriber.next(connected ? AppState.APP_OPEN : AppState.APP_CLOSED)
+    }, 500)
+})
+
+const subject = new Subject<AppState>()
+observable.subscribe(subject)
+
+const sendMessage = sendApduMsg.bind(null, CLA, handleError)
+
+const generateGetPublicKeyResponse = parseResponse.bind(null, response =>
+    ({
+        publicKey: response.slice(0, 33),
+    }),
+)
+
+const generateSignResponse = parseResponse.bind(null, response =>
+    ({
+        signature: response.slice(0, 64),
+    }),
+)
+
+const generateGetVersionResponse = parseResponse.bind(null, response =>
+    ({
+        CLA: response.slice(0, 1),
+        locked: response.slice(4, 5),
+        version: `
+            ${parseInt(response.slice(1, 2).toString('hex'), 16)}.
+            ${parseInt(response.slice(2, 3).toString('hex'), 16)}.
+            ${parseInt(response.slice(3, 4).toString('hex'), 16)}
+        `,
+    }))
+
+const getPublicKey = (bip44: string, p1: 0 | 1 = 0): Promise<{ publicKey: Buffer }> =>
+    sendMessage(
+        generateGetPublicKeyResponse,
+        Instruction.INS_GET_PUBLIC_KEY,
+        Buffer.from(bip44, 'hex'),
+        p1,
+    )
+
+const signHash = (bip44: string, hash: Buffer): Promise<{ signature: Buffer }> =>
+    sendMessage(
+        generateSignResponse,
+        Instruction.INS_SIGN_HASH,
+        Buffer.concat([Buffer.from(bip44, 'hex'), hash]),
+        20,
+        hash.length,
+    )
+
+const getVersion = () =>
+    sendMessage(
+        generateGetVersionResponse,
+        Instruction.INS_GET_VERSION,
+    )
+
+const getVersionPublic = async () => {
+    await versionPromise
+    return version
+}
+
```

Comment:
I preferred to keep the big signAtom method without shared state, and made a separate method for that.

The reason for the isSigning state is because the ledger app crashes if a message is sent to it while it's already processing another message. And, the signAtom method needs to process several messages before it's done.

Furthermore, I'm constantly polling the ledger for it's status by sending a message.

So in order it prevent the ledger app from crashing while it's signing, I'm enabling this check so that the polling can wait until the signing is over.